### PR TITLE
Add deep sleep support

### PR DESCRIPTION
**Pull request by zhaofengli**
This PR implements deep sleep (suspend-to-RAM) on Trimui Brick and Smart Pro. The device enters deep sleep instead of powering off after 2 minutes of faux sleep, making resumption much faster (~1s) than a cold boot (~15s) with negligible battery drain (~3% per day). The stock UI on those two devices also uses deep sleep so this should be safe.

The 2-minute faux sleep is preserved to make the UX snappy for short pauses. To immediately enter deep sleep after pressing the power button, revert the "all: Enter deep sleep after 2 minutes of faux sleep" commit.

### CODE REVIEW

**Comment by zhaofengli** [comment]
tl;dr: I added a shell script to perform the necessary teardown/setup on the suspend/resume process. Without the script, the device may fail to suspend if Wi-Fi is configured but not connected[^wifi], and sound can stop working occasionally on resume[^sound]. So far, deep sleep has been working well in my testing.

[^wifi]: See commit "tg3040, tg5040: Save and restore the alsa mixer state on suspend"
[^sound]: See commit "tg3040, tg5040: Add suspend script"

---

I took my Trimui Brick outside today. It has Wi-Fi set up (unsupported, I know) and I encountered this annoying behavior: If any network is configured but none is in-range, suspending can fail while the driver performs a scan:

```
[   67.068447] [PM_WRN] Don't suspend because of pm_state need keep awake.
[   67.068484] dpm_run_callback(): wiphy_suspend+0x0/0xbc returns -16
[   67.068490] PM: Device phy0 failed to suspend: error -16
[   67.068494] PM: Some devices failed to suspend, or early wake event detected
[   67.068523] PM: resume of devices complete after 0.022 msecs
[   67.104539] PM: Finishing wakeup.
[   67.108257] Restarting tasks ... done.
[ERROR] failed to set power state: 16
```

In this case, MinUI falls back to powering off which is the status quo, but it would be nice to get deep sleep to work reliably. My finding so far:

- ✅ No network is configured: Can suspend regardless of whether wpa_supplicant is running
- ✅ Connected: Can suspend
- ❌ Network configured but none in-range: Cannot suspend with wpa_supplicant running

The stock keymon stops wpa_supplicant among other processes, and I put this logic inside a shell script so it's easy for advanced users to customize. With this script, deep sleep works reliably from my (albeit limited) testing.

---

Now MinUI will make 3 attempts to enter deep sleep before falling back to powering off. The system remains in faux sleep and can be immediately woken up between attempts, so it's all transparent to the user.

If you don't want the Wi-Fi teardown logic in the repo, we can just drop the "tg3040, tg5040: Add suspend script" commit and have users supply the script themselves. I have tested that deep sleep does work reliably if no Wi-Fi network is configured. In any case, the system will fall back to powering off so there is no change in behavior if deep sleep fails.

**Comment by zhaofengli** [comment]
In case people want to test, I built some binaries attached below. Note that I don't recommend downloading random binaries in a GitHub thread - Would be nice if this repo had a CI that automatically built everything.

[minui-deep-sleep-bin-dist.zip](https://github.com/user-attachments/files/18265884/minui-deep-sleep-bin-dist.zip)

- Brick: Copy files in `tg3040` to `.system/tg3040/bin`
- Smart Pro: Copy files in `tg5040` to `.system/tg5040/bin`

To return to official MinUI, download the latest release then replace `minui.elf` and `minarch.elf`.

**Comment by shauninman** [comment]
This changes core functionality and breaks consistency in behavior across supported devices. So while I appreciate the effort that went into this I’m not going to accept it at this time (maybe I’ll revisit when more supported devices have acceptable native sleep performance).

**Comment by zhaofengli** [comment]
Hi, fair enough - MinUI doesn't have to be everything for everyone. At the same time, it's already specialized for each device and I think supporting basic hardware features improves UX without breaking consistency. I often found myself pressing the power button and realizing I had left it sleeping for too long and had to long-press to turn it back on again.

By the way, I really like MinUI and appreciate the restraint in adding features in general. The simplicity is what makes it so delightful to use.

**Comment by LijMeh** [comment]
@zhaofengli quick question: would this code be extendable to other devices that support deep sleep, or is it pretty specific to the TrimUI devices? (in my case, for a custom fork for my Miyoo Flip)

**Comment by zhaofengli** [comment]
@LijMeh It's extensible and you can look at the PR commit-by-commit to see how I approached it. In short, this is what you need to do in your device:

1. [Make `PLAT_supportsDeepSleep` return 1.](https://github.com/shauninman/MinUI/pull/24/commits/cad5a469b023457817533d79c26658ff824fc26f)
    - This will enable the [default `PLAT_deepSleep` implementation](https://github.com/shauninman/MinUI/pull/24/commits/8266da93b05d5534d45ce538560ecbaa8dad98a6#diff-a3338877d525fd79e61d088a0a378385a24d7a2e4885ae5aec4c4772f9712dd2R1350-R1371) that simply enters suspend-to-RAM using [the standard sysfs interface](https://www.kernel.org/doc/Documentation/power/states.txt) without special preparation.
    - For Trimui, this already works except in the cases I mentioned [above](https://github.com/shauninman/MinUI/pull/24#issuecomment-2563883541).
3. (Optional) If your device needs special handling, override `PLAT_deepSleep`.
4. (Optional) If your device needs even more complex handling, implement it as a shell script and drop it at `(sdcard)/.system/{device}/bin/suspend`. This shell script will be [run in place of `PLAT_deepSleep`](https://github.com/shauninman/MinUI/pull/24/commits/b2a635def7087ece4a637dbb90c45124a717d932).
    - This is what I settled upon for [my Trimui devices](https://github.com/shauninman/MinUI/pull/24/commits/9318f31453bf3d2575d255fa0b4ff5cec2dd08c1)

**Comment by LijMeh** [comment]
@zhaofengli Thanks for the info! I’ll give it a shot.

**Comment by LijMeh** [comment]
@zhaofengli got it working, and with a bit of code to respond to the lid status it seems perfect. Only small issue is the volume bar shows up after waking and won’t hide until a volume button is pressed. Did you see anything like that on the TrimUI? (If not, happy to troubleshoot myself, not asking you to fix it lol). 

Was going to share with other Flip users, will credit your GitHub unless there’s any reason you’d prefer I not?

**Comment by zhaofengli** [comment]
@LijMeh Nice! I don't see the behavior regarding volume button, but I did observe MinUI recognizing spurious power button presses after waking up (the pressed-down state seems to persist through the sleep). I addressed this by [ignoring power button presses for a short period](https://github.com/shauninman/MinUI/pull/24/commits/8266da93b05d5534d45ce538560ecbaa8dad98a6#diff-a3338877d525fd79e61d088a0a378385a24d7a2e4885ae5aec4c4772f9712dd2R1540-R1545) after resuming. What you are seeing with the volume button could be similar, and logging the button presses can help diagnose the issue.

I'm happy for you to share it, just one request - Please merge/cherry-pick so the commit history is preserved. I see that you already have a [branch](https://github.com/LijMeh/MinUI/commit/851e15ea3eb47a8ce996813abc87daf64a7ce07e) with copy-pasted code, and rebasing it on top of this PR is the easiest way to keep the history. See the [git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase) tutorial (it's a very powerful and useful tool in general).

**Comment by LijMeh** [comment]
@zhaofengli Sounds good, I'll make sure to redo the branch with a merge before I'm ready to release anything. Can't figure out that volume bug at the moment so not planning to share until I do. 

(I did notice the sleep issue and was able to fix that, but I think this might be related to something from the base firmware, as it also displays the volume after sleep in the Miyoo software, and everything I've tried to do to clear the volume within MinUI hasn't worked).
